sdk/communication/Azure.Communication.Administration/src/ReleasePhoneNumberOperation.cs

```diff
@@ -0,0 +1,139 @@
+
+        /// <inheritdocs />
+        public override bool HasCompleted => _hasCompleted;
+
+        /// <inheritdocs />
+        public override bool HasValue => _value != null;
+
+        /// <inheritdocs />
+        public override Response GetRawResponse() => _rawResponse;
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        public override Response UpdateStatus(
+            CancellationToken cancellationToken = default) =>
+            UpdateStatusAsync(false, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        public override async ValueTask<Response> UpdateStatusAsync(
+            CancellationToken cancellationToken = default) =>
+            await UpdateStatusAsync(true, cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <param name="async" />
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        private async Task<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            if (HasCompleted)
+            {
+                return GetRawResponse();
+            }
+
+            if (cancellationToken == default)
+            {
+                cancellationToken = _cancellationToken;
+            }
+
+            Response<PhoneNumberRelease> update = async
+                ? await _client.GetReleaseByIdAsync(releaseId: Id, cancellationToken: cancellationToken).ConfigureAwait(false)
+                : _client.GetReleaseById(releaseId: Id, cancellationToken: cancellationToken);
+
+            _value = update.Value;
+
+            var terminateStatuses = new ReleaseStatus[]
+            {
+                ReleaseStatus.Complete,
+                ReleaseStatus.Failed,
+                ReleaseStatus.Expired
+            };
+
+            if (update.Value.Status.HasValue)
+            {
+                if (terminateStatuses.Contains(update.Value.Status.Value))
+                {
+                    _hasCompleted = true;
```

Comment:
Just noticed this is not thread safe, there is a race condition when one thread gets a terminate status and another thread does not.

Reply:
Perhaps we should simplify `HasCompleted` to something like this:

```C#
public override bool HasCompleted => terminateStatuses.Contains(_value?.Status?.Value)
```

Reply:
[setting `_rawResponse`](https://github.com/Azure/azure-sdk-for-net/pull/15803/files#diff-b68e744e8d36a7d83eca38d48616eaa44dafb59573e9a0a69ff83f9a4fa541b1R127) could also be affected by a race condition

Reply:
@christothes are the operations supposed to be thread safe or not? I think other operations also have similar timing issues. We are caching `rawResponse` and `isCompleted`, but technically a thread that was running a bit behind can overwrite the cached response, while another thread set the `isCompleted` to true.
https://github.com/Azure/azure-sdk-for-net/blob/d60d77db6ea287478fdd22c4b13fd1c5900be86f/sdk/keyvault/Azure.Security.KeyVault.Certificates/src/CertificateOperation.cs#L133
https://github.com/Azure/azure-sdk-for-net/blob/a7e300557be4370b3bdedda3fcb42ef846e3fb44/sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeReceiptsOperation.cs#L179

Reply:
Discussed offline - although rare, it does seem like there could be a threading issue here. We can add a lock to be sure about it.

Reply:
I was about to disagree because the `WaitForCompletionAsync` awaits the update operation before sending a new one. But then I realized that `UpdateStatus` is also public on the Operation, so the scheduled update is theoretically in a race condition with the manual one.